### ISSUE TITLE
FSE enqueue scripts for editor iframe [.editor-styles-wrapper]

### ISSUE THREAD

**Issue by amir2mi** [WordPress/gutenberg#34763]
I'm searching for about a week when I started working with the FSE feature.
At the first sight, FSE was different than a regular post because the preview is in the **iframe** and it makes things e.g custom blocks difficult to be the same both in FSE and a regular post.
One of these difficulties is enqueuing scripts for the editor inside that iframe.
At this moment, it is possible to enqueue assets both with `block.json` and `enqueue_block_assets` hook but as I tried `enqueue_block_assets`, it does not enqueue **scripts** inside that iframe, CSS is just fine but scripts no, they are enqueued outside of the `.editor-styles-wrapper`.
I also tried the `enqueue_block_editor_assets` hook which did something similar.
This is the problem: **there is no option to enqueue a script inside the `.editor-styles-wrapper` iframe just for editor**.

**Comment by annezazu** [comment]
@ellatrix perhaps you have some thoughts to share here!

**Comment by scrobbleme** [comment]
This should be fixed...

As a workaround, have a look into WooCommerce code, they have [workaround](https://github.com/woocommerce/woocommerce-gutenberg-products-block/blob/271f04dde620fa2254ab5e72ba3ca0bec5da52a5/src/AssetsController.php#L135):

Basically:

```

add_action( 'enqueue_block_editor_assets', 'my_enqueue_block_editor_assets' );
add_action( 'admin_enqueue_scripts', 'my_update_block_style_dependencies' , 20 );


function my_enqueue_block_editor_assets() {
    wp_enqueue_style( 'my-style',  ...);
}

/** Thanks to WooCommerce     */
public function my_update_block_style_dependencies() {
    $wp_styles = wp_styles();
    $style     = $wp_styles->query( 'wc-blocks-style', 'registered' );

    if (  $style ) {
	$style->deps[] = 'my-style';
   }
}
```

**Comment by leeshadle** [comment]
Thank you for creating this issue @amir2mi! I am seeing the same thing...

When using enqueue_block_editor_assets for enqueuing scripts in the iframed 'Full Site Editor' they get loaded outside of the iframed 'Full Site Editor', whereas styles get loaded inside the iframed 'Full Site Editor'.

This seems like unexpected behavior and I'd love access and the ability to add scripts within the iframed 'Full Site Editor' :)

In non-iframed editors ( pages, posts, custom post types, etc. ) enqueueing scripts gives plugin/theme developers access to the document the editor is loaded in.

For example, I created controls that allows you to change a block-based theme's contentSize and wideSize by dynamically appending css properties to the document.  The css properties get appended to the document in a way that they can be consumed in the non-iframed editor, however they get appended to the outer document that contains the iframed 'Full Site Editor', not the iframed editor itself.

**Comment by pryley** [comment]
This sort of works:

```php
/**
 * @param array $settings
 * @param \WP_Block_Editor_Context $context
 * @return array
 */
add_filter('block_editor_settings_all', function ($settings, $context) {
    global $pagenow;
    if ('site-editor.php' === $pagenow) {
        // wp_enqueue_script using $handle
        $handles = [$handle];
        $done = wp_scripts()->done;
        ob_start();
        wp_scripts()->done = [];
        wp_scripts()->do_items($handles);
        wp_scripts()->done = $done;
        $scripts = ob_get_clean();
        $settings['__unstableResolvedAssets']['scripts'] .= $scripts;
    }
    return $settings;
}, 10, 2);
```

**Comment by ellatrix** [comment]
This is a duplicate of #44945, which was fixed by #49655. See https://make.wordpress.org/core/2023/07/18/miscellaneous-editor-changes-in-wordpress-6-3/#post-editor-iframed for more information on how to enqueue assets in the iframe.
